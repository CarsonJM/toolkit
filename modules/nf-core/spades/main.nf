process SPADES {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/spades:4.0.0--h5fb382e_1' :
        'biocontainers/spades:4.0.0--h5fb382e_1' }"

    input:
    tuple val(meta), path(illumina), path(pacbio), path(nanopore)
    path yml
    path hmm

    output:
    tuple val(meta), path('*.scaffolds.fa.gz')    , optional:true, emit: scaffolds
    tuple val(meta), path('*.contigs.fa.gz')      , optional:true, emit: contigs
    tuple val(meta), path('*.transcripts.fa.gz')  , optional:true, emit: transcripts
    tuple val(meta), path('*.gene_clusters.fa.gz'), optional:true, emit: gene_clusters
    tuple val(meta), path('*.assembly.gfa.gz')    , optional:true, emit: gfa
    tuple val(meta), path('warnings.log')         , optional:true, emit: warnings
    tuple val(meta), path('*.spades.log')         , emit: log
    tuple val(meta), env(min_kmer)                , emit: min_kmer
    tuple val(meta), env(max_kmer)                , emit: max_kmer
    path  "versions.yml"                          , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def maxmem = task.memory.toGiga()
    def illumina_reads = illumina ? ( meta.single_end ? "-s $illumina" : "-1 ${illumina[0]} -2 ${illumina[1]}" ) : ""
    def pacbio_reads = pacbio ? "--pacbio $pacbio" : ""
    def nanopore_reads = nanopore ? "--nanopore $nanopore" : ""
    def custom_hmms = hmm ? "--custom-hmms $hmm" : ""
    def reads = yml ? "--dataset $yml" : "$illumina_reads $pacbio_reads $nanopore_reads"
    """
    spades.py \\
        $args \\
        --threads $task.cpus \\
        --memory $maxmem \\
        $custom_hmms \\
        $reads \\
        -o ./
    mv spades.log ${prefix}.spades.log

    if [ -f scaffolds.fasta ]; then
        mv scaffolds.fasta ${prefix}.scaffolds.fa
        gzip -n ${prefix}.scaffolds.fa
    fi
    if [ -f contigs.fasta ]; then
        mv contigs.fasta ${prefix}.contigs.fa
        gzip -n ${prefix}.contigs.fa
    fi
    if [ -f transcripts.fasta ]; then
        mv transcripts.fasta ${prefix}.transcripts.fa
        gzip -n ${prefix}.transcripts.fa
    fi
    if [ -f assembly_graph_with_scaffolds.gfa ]; then
        mv assembly_graph_with_scaffolds.gfa ${prefix}.assembly.gfa
        gzip -n ${prefix}.assembly.gfa
    fi

    if [ -f gene_clusters.fasta ]; then
        mv gene_clusters.fasta ${prefix}.gene_clusters.fa
        gzip -n ${prefix}.gene_clusters.fa
    fi

    # identify min/max kmer size
    kmer_string=\$(grep "K values to be used: \\[" ${prefix}.spades.log | sed 's/.*K values to be used: \\[//; s/\\].*//')
    kmer_array=(\${kmer_string//,/ })
    min_kmer=\$(IFS=\$'\\n'; echo "\${kmer_array[*]}" | sort -nr | tail -n 1)
    max_kmer=\$(IFS=\$'\\n'; echo "\${kmer_array[*]}" | sort -nr | head -n 1)
    echo \$min_kmer
    echo \$max_kmer

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spades: \$(spades.py --version 2>&1 | sed -n 's/^.*SPAdes genome assembler v//p')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def maxmem = task.memory.toGiga()
    def illumina_reads = illumina ? ( meta.single_end ? "-s $illumina" : "-1 ${illumina[0]} -2 ${illumina[1]}" ) : ""
    def pacbio_reads = pacbio ? "--pacbio $pacbio" : ""
    def nanopore_reads = nanopore ? "--nanopore $nanopore" : ""
    def custom_hmms = hmm ? "--custom-hmms $hmm" : ""
    def reads = yml ? "--dataset $yml" : "$illumina_reads $pacbio_reads $nanopore_reads"
    """
    echo "" | gzip > ${prefix}.scaffolds.fa.gz
    echo "" | gzip > ${prefix}.contigs.fa.gz
    echo "" | gzip > ${prefix}.transcripts.fa.gz
    echo "" | gzip > ${prefix}.gene_clusters.fa.gz
    echo "" | gzip > ${prefix}.assembly.gfa.gz
    touch ${prefix}.spades.log
    touch warnings.log
    min_kmer=21
    max_kmer=51

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spades: \$(spades.py --version 2>&1 | sed -n 's/^.*SPAdes genome assembler v//p')
    END_VERSIONS
    """
}
