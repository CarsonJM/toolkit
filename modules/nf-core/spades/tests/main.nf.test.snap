{
    "sarscov2 - pe - nanopore ": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.scaffolds.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.contigs.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.assembly.gfa.gz:md5,19418df83534fc93543dec4ec9b2ae72"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "73"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:19:03.165250013"
    },
    "sarscov2 - pe - hmm ": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.scaffolds.fa.gz:md5,6263e4b88b3e0099ab0abb0bd034a581"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.assembly.gfa.gz:md5,07136eab8e231f095dc5dd62f1b62a91"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "123"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:22:28.756834169"
    },
    "sarscov2 - pe - pacbio ": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.scaffolds.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.contigs.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.assembly.gfa.gz:md5,19418df83534fc93543dec4ec9b2ae72"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "73"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:18:41.943915484"
    },
    "sarscov2 - pe ": {
        "content": [
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.contigs.fa.gz:md5,70e4a5485dd59566b212a199c31c343b"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.assembly.gfa.gz:md5,b773132d52be5090cdbdf5a643027093"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "73"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:18:19.017461119"
    },
    "sarscov2 - pe - nanopore - yml ": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.scaffolds.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.contigs.fa.gz:md5,7ddaf03740df422a93fcaffbcd7e9679"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.assembly.gfa.gz:md5,19418df83534fc93543dec4ec9b2ae72"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "73"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:19:23.611405729"
    },
    "sarscov2 - se ": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.scaffolds.fa.gz:md5,65ba6a517c152dbe219bf4b5b92bdad7"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.contigs.fa.gz:md5,65ba6a517c152dbe219bf4b5b92bdad7"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.assembly.gfa.gz:md5,e4836fdf7104d79e314e3e50986b4bb2"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "21"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "73"
                ]
            ],
            [
                "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:17:58.027321834"
    },
    "sarscov2 - pe - stub ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.scaffolds.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.contigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.transcripts.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gene_clusters.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.assembly.gfa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "warnings.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.spades.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "21"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "51"
                    ]
                ],
                "9": [
                    "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
                ],
                "contigs": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.contigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "gene_clusters": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gene_clusters.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "gfa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.assembly.gfa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.spades.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "max_kmer": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "51"
                    ]
                ],
                "min_kmer": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "21"
                    ]
                ],
                "scaffolds": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.scaffolds.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "transcripts": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.transcripts.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,990abcdf543421412170e5cf413ec56d"
                ],
                "warnings": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "warnings.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:20:01.763977329"
    }
}